src/surrogate_heat_driver.cpp

```diff
@@ -380,21 +384,368 @@ std::vector<double> SurrogateHeatDriver::volume() const
 
-  // Determine indices
-  gsl::index pin = local_elem / (n_axial_ * n_rings() * n_azimuthal_);
-  gsl::index axial = (local_elem / (n_rings() * n_azimuthal_)) % n_axial_;
-  gsl::index ring = (local_elem / n_azimuthal_) % n_rings();
-  gsl::index azimuthal = local_elem % n_azimuthal_;
+  // get assembly index:
+  // calculated index is relative to active assemblies, so need to account
+  // for any skipped assemblies in the full core. Add 1 for each skipped assembly.
+  gsl::index assem = (local_elem / (n_pins_ * n_axial_ * n_rings() * n_azimuthal_));
+  if (n_skip_ > 0) {
+    for (gsl::index i = 0; i < skip_assemblies_.size(); ++i) {
+      if (skip_assemblies_[i] <= assem) {
+        assem = assem + 1;
+      }
+    }
+  }
+
+  if (assembly_drivers_[assem].skip_assembly_){
+    return 0;
+  }
+
+  // Determine indices within assembly
+  gsl::index assem_local_elem =
+    local_elem % (n_pins_ * n_axial_ * n_rings() * n_azimuthal_);
+  gsl::index pin = assem_local_elem / (n_axial_ * n_rings() * n_azimuthal_);
+  gsl::index axial = (assem_local_elem / (n_rings() * n_azimuthal_)) % n_axial_;
+  gsl::index ring = (assem_local_elem / n_azimuthal_) % n_rings();
+  gsl::index azimuthal = assem_local_elem % n_azimuthal_;
 
   // Set heat source
-  source_(pin, axial, ring, azimuthal) = heat;
+  assembly_drivers_[assem].source_(pin, axial, ring, azimuthal) = heat;
   return 0;
 }
 
-double SurrogateHeatDriver::rod_axial_node_power(const int pin, const int axial) const
+void SurrogateHeatDriver::solve_step()
+{
+  timer_solve_step.start();
+  if (has_coupling_data()) {
+    // iterate over each assembly for each solve step
+    for (gsl::index row = 0; row < n_assem_y_; ++row) {
+      for (gsl::index col = 0; col < n_assem_x_; ++col) {
+        int assem_index = row * n_assem_x_ + col;
+        if (!assembly_drivers_[assem_index].skip_assembly_) {
+          comm_.message("Solving fluid equation for assembly " +
+                        std::to_string(assem_index) + " ...");
+          assembly_drivers_[assem_index].solve_fluid();
+          comm_.message("Solving heat equation for assembly " +
+                        std::to_string(assem_index) + " ...");
+          assembly_drivers_[assem_index].solve_heat();
+        }
+      }
+    }
+  }
+  timer_solve_step.stop();
+}
+
+void SurrogateHeatDriver::write_step(int timestep, int iteration)
+{
+  // TODO: Timers deadlock here...
+  // timer_write_step.start();
+  if (!has_coupling_data())
+    return;
+
+  // if called, but viz isn't requested for the situation,
+  // exit early - no output
+  if ((iteration < 0 && "final" != viz_iterations_) ||
+      (iteration >= 0 && "all" != viz_iterations_)) {
+    return;
+  }
+
+  // otherwise construct an appropriate filename and write the data
+  std::stringstream filename_base;
+  filename_base << viz_basename_;
+  if (iteration >= 0 && timestep >= 0) {
+    filename_base << "_t" << timestep << "_i" << iteration;
+  }
+
+  // write one file per assembly
```

Comment:
Can you elaborate a little on the choice to write one file per assembly? Writing a single file might be more convenient for a user but I can see that there would be some potential issues.

Reply:
It required the least amount of changes to the current code to write one file per assembly. I haven't quite figured out how to make it one file, though I admittedly haven't thought too deeply on it yet. But I can see what I can do to make that happen.